clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/customtemplates/CustomTemplate.kt

```diff
@@ -0,0 +1,271 @@
+
+        override val thisRef: FunctionBuilder = this
+    }
+
+    /**
+     * Builder for [CustomTemplate] code templates. "." characters in template arguments' names denote hierarchical
+     * structure. They are treated the same way as the keys provided by [mapArgument]. If a higher-level name (to the left
+     * of a . symbol) matches a map argument's name it is treated the same as if the argument was part of the map itself.
+     *
+     * For example, the following code snippets define identical arguments:
+     * ```
+     * builder.mapArgument(
+     *     name = "map",
+     *     value = mapOf (
+     *         "a" to 5,
+     *         "b" to 6
+     *     )
+     * )
+     * ```
+     * and
+     * ```
+     * builder.intArgument("map.a", 5)
+     * builder.intArgument("map.b", 6)
+     * ```
+     */
+    class TemplateBuilder : Builder<TemplateBuilder>(TEMPLATE, allowHierarchicalNames = true, isVisual = true) {
+
+        override val thisRef: TemplateBuilder = this
+
+        /**
+         * Add a map structure to the arguments of the [CustomTemplate]. The [name] should be unique across all
+         * arguments and also all keys in [value] should form unique names across all arguments.
+         *
+         * @param value The map must be non-empty. Values can be of type [Byte], [Short], [Int], [Long], [Float],
+         * [Double], [Boolean], [String] or another [Map]<String, Any> which values can also be of the same types.
+         */
+        fun mapArgument(name: String, value: Map<String, Any>): TemplateBuilder {
+            if (value.isEmpty()) {
+                throw CustomTemplateException("Map argument must not be empty")
+            }
+
+            for (mapEntry in value) {
+                val argValue = mapEntry.value
+                val argName = "$name.${mapEntry.key}"
+
+                @Suppress("UNCHECKED_CAST")
+                when (argValue) {
+                    is Byte -> byteArgument(argName, argValue)
+                    is Short -> shortArgument(argName, argValue)
+                    is Int -> intArgument(argName, argValue)
+                    is Long -> longArgument(argName, argValue)
+                    is Float -> floatArgument(argName, argValue)
+                    is Double -> doubleArgument(argName, argValue)
+                    is Boolean -> booleanArgument(argName, argValue)
+                    is String -> stringArgument(argName, argValue)
+                    is Map<*, *> -> mapArgument(argName, argValue as Map<String, Any>)
+                    else -> throw CustomTemplateException("Unsupported value type ${argValue.javaClass} for argument $argName")
+                }
+            }
+            return this
+        }
+
+        fun actionArgument(name: String): TemplateBuilder {
+            addArgument(name, ACTION, null)
+            return this
+        }
+    }
+
+    sealed class Builder<T : Builder<T>>(
```

Comment:
how do we handle thrown exceptions? will this cause crash in client or we will catch it internally?

Reply:
We've designed the API around defining/registering templates to fail fast. We decided on that with these arguments in mind: 1. Template definitions with invalid state (names, argNames, types) are useless - they would either be rejected from the backend or not work as intended when trying to use them later. 2. Template definitions are somewhat declaratively defined and raising an exception would immediately signal for the bad definition and the place it occurs instead of having default values or ignoring bad values (which could result in unexpected behavior for the user). 3. The templates are supposed to be registered around app launch and we assume that developers would run their app at least once when defining templates. Also the templates should be synched with the backend in "dev mode" and that is at least one run. 

Once the templates are successfully registered we would not throw exceptions in order not to crash the clients apps as is with the other APIs in the SDK.